nearcache.go

```diff
@@ -0,0 +1,690 @@
+		rc = adapter
+		se = adapter
+	} else {
+		adapter := nearCacheValueStoreAdapter{ss: ss}
+		rc = adapter
+		se = adapter
+	}
+	return &nearCache{
+		cfg:   cfg,
+		store: newNearCacheRecordStore(cfg, ss, rc, se),
+	}
+}
+
+func (nc *nearCache) Get(key interface{}) (interface{}, bool, error) {
+	_, ok := key.(serialization.Data)
+	if nc.cfg.SerializeKeys {
+		if !ok {
+			panic("key must be of type serialization.Data!")
+		}
+	} else if ok {
+		panic("key cannot be of type Data!")
+	}
+	return nc.store.Get(key)
+}
+
+func (nc *nearCache) Invalidate(key interface{}) {
+	nc.store.Invalidate(key)
+}
+
+func (nc nearCache) Size() int {
+	return nc.store.Size()
+}
+
+func (nc nearCache) Stats() nearcache.Stats {
+	return nc.store.Stats()
+}
+
+func (nc nearCache) InvalidationRequests() int64 {
+	return nc.store.InvalidationRequests()
+}
+
+func (nc *nearCache) TryReserveForUpdate(key interface{}, keyData serialization.Data, ups nearCacheUpdateSemantic) (int64, error) {
+	// eviction stuff will be implemented in another PR
+	// nearCacheRecordStore.doEviction(false);
+	return nc.store.TryReserveForUpdate(key, keyData, ups)
+}
+
+func (nc *nearCache) TryPublishReserved(key, value interface{}, reservationID int64) (interface{}, error) {
+	cached, err := nc.store.TryPublishReserved(key, value, reservationID, true)
+	if err != nil {
+		return nil, err
+	}
+	if cached != nil {
+		value = cached
+	}
+	return value, nil
+}
+
+const (
+	nearCacheRecordStoreTimeNotSet int64 = -1
+	pointerCostInBytes                   = (32 << uintptr(^uintptr(0)>>63)) >> 3
+	int32CostInBytes                     = 4
+	int64CostInBytes                     = 8
+	atomicValueCostInBytes               = 8
+	uuidCostInBytes                      = 16 // low uint64 + high uint64
+	nearCacheRecordNotReserved     int64 = -1
+	nearCacheRecordReadPermitted         = -2
+)
+
+type nearCacheRecordValueConverter interface {
+	ConvertValue(value interface{}) (interface{}, error)
+}
+
+type nearCacheStorageEstimator interface {
+	GetRecordStorageMemoryCost(rec *nearCacheRecord) int64
+}
+
+type nearCacheDataStoreAdapter struct {
+	ss *serialization.Service
+}
+
+func (n nearCacheDataStoreAdapter) ConvertValue(value interface{}) (interface{}, error) {
+	if value == nil {
+		// have to check value manually here,
+		// otherwise n.ss.ToData returns type + nil, which is not recognized as nil
+		return nil, nil
+	}
+	return n.ss.ToData(value)
+}
+
+func (n nearCacheDataStoreAdapter) GetRecordStorageMemoryCost(rec *nearCacheRecord) int64 {
+	if rec == nil {
+		return 0
+	}
+	cost := pointerCostInBytes + // the record is stored as a pointer in the map
+		5*int64CostInBytes + // CreationTime, lastAccessTime, ExpirationTime, InvalidationSequence, reservationID
+		3*int32CostInBytes + // PartitionID, hits, cachedAsNil
+		1*atomicValueCostInBytes + // holder of the value
+		1*uuidCostInBytes // UUID
+	value := rec.Value()
+	data, ok := value.(serialization.Data)
+	if ok {
+		cost += data.DataSize()
+	}
```

Comment:
It seems the value of the record can be nil ATM, since this method can be called earlier than the Java implementation so the if (or if check against nil) is necessary. Let me check that...

Reply:
value should be either nil or serialization.Data since `nearCacheDataStoreAdapter.ConvertValue`

Reply:
Removed that check at: https://github.com/hazelcast/hazelcast-go-client/pull/806/commits/5e33d6ad56495f39428a579a54d530864570ac49